pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -304,77 +322,29 @@ func (a Adapter) createOrUpdateComponent(componentExists bool, ei envinfo.EnvSpe
-	if err != nil {
-		return err
-	}
-
-	if !pref.GetEphemeralSourceVolume() {
-
-		// If ephemeral volume is false, then we need to add to source volume in the map to create pvc.
-		containerNameToVolumes["odosource"] = []common.DevfileVolume{
-			{
-				Name: utils.OdoSourceVolume,
-				Size: utils.OdoSourceVolumeSize,
-			},
-		}
-	}
-
 	var odoSourcePVCName string
 
-	var uniqueStorages []common.Storage
 	volumeNameToPVCName := make(map[string]string)
-	processedVolumes := make(map[string]bool)
-
-	// Get a list of all the unique volume names and generate their PVC names
-	// we do not use the volume components which are unique here because
-	// not all volume components maybe referenced by a container component.
-	// We only want to create PVCs which are going to be used by a container
-	for _, volumes := range containerNameToVolumes {
-		for _, vol := range volumes {
-			if _, ok := processedVolumes[vol.Name]; !ok {
-				processedVolumes[vol.Name] = true
-
-				// Generate the PVC Names
-				klog.V(2).Infof("Generating PVC name for %v", vol.Name)
-				generatedPVCName, err := storage.GeneratePVCNameFromDevfileVol(vol.Name, componentName)
-				if err != nil {
-					return err
-				}
 
-				// Check if we have an existing PVC with the labels, overwrite the generated name with the existing name if present
-				existingPVCName, err := storage.GetExistingPVC(&a.Client, vol.Name, componentName)
-				if err != nil {
-					return err
-				}
-				if len(existingPVCName) > 0 {
-					klog.V(2).Infof("Found an existing PVC for %v, PVC %v will be re-used", vol.Name, existingPVCName)
-					generatedPVCName = existingPVCName
-				}
-
-				if vol.Name == utils.OdoSourceVolume {
-					odoSourcePVCName = generatedPVCName
-				}
-
-				pvc := common.Storage{
-					Name:   generatedPVCName,
-					Volume: vol,
-				}
-				uniqueStorages = append(uniqueStorages, pvc)
-				volumeNameToPVCName[vol.Name] = generatedPVCName
-			}
-		}
+	// handle the ephemeral storage
+	err = utils.HandleEphemeralStorage(a.Client, storageClient, a.ComponentName)
+	if err != nil {
+		return err
 	}
 
-	err = storage.DeleteOldPVCs(&a.Client, componentName, processedVolumes)
+	// list all the pvcs for the component
+	pvcs, err := a.Client.ListPVCs(fmt.Sprintf("%v=%v", componentlabels.ComponentLabel, a.ComponentName))
 	if err != nil {
 		return err
 	}
 
-	// remove odo source volume from these maps as we do not want to pass source volume in GetPVCAndVolumeMount
-	// we are mounting odo source volume seperately
-	delete(volumeNameToPVCName, utils.OdoSourceVolume)
-	delete(containerNameToVolumes, "odosource")
+	for _, pvc := range pvcs {
+		// check if the pvc is in the terminating state
+		if pvc.DeletionTimestamp != nil {
+			continue
+		}
+		volumeNameToPVCName[pvc.Labels[storagelabels.DevfileStorageLabel]] = pvc.Name
+	}
 
 	// Get PVC volumes and Volume Mounts
 	containers, pvcVolumes, err := storage.GetPVCAndVolumeMount(containers, volumeNameToPVCName, containerNameToVolumes)
```

Comment:
so i was looking to extract this logic out, so that the library can return the updated containers with volume mounts and the pvc volumes after reading the devfile.

Reply:
I have only modified the network calls in this function.

Reply:
Yep, after this merged, I can look to consolidate this in devfile/library call; so anyone calling devfile/library generator func like `generator.GetVolume()` will return the `pvcVolumes` and updated container volumeMounts 👍